### PR TITLE
Feature/new firmware retry on start reply timeout

### PR DESCRIPTION
**Pull request by rfeistenauer**
Adds feature: sending start request again when waiting for reply timed out

### CODE REVIEW

File: include/psen_scan_v2/scanner_controller.h [hunk 1]
```diff
@@ -124,11 +126,15 @@ ScannerControllerT<TCSM, TUCI>::ScannerControllerT(const ScannerConfiguration& s
 }
 
 template <typename TCSM, typename TUCI>
-void ScannerControllerT<TCSM, TUCI>::handleNewMonitoringFrame(const MaxSizeRawData& data, const std::size_t& num_bytes)
+void ScannerControllerT<TCSM, TUCI>::handleUdpData(const MaxSizeRawData& data, const std::size_t& num_bytes)
 {
-  MonitoringFrameMsg frame{ MonitoringFrameMsg::deserialize(data, num_bytes) };
-  state_machine_.processMonitoringFrameReceivedEvent();
+  MonitoringFrameMsg frame{ MonitoringFrameMsg::fromRawData(data, num_bytes) };
```
Comment (SansoneG) [inline include/psen_scan_v2/scanner_controller.h:131]: `fromRawData` does not exist.

File: include/psen_scan_v2/scanner_controller.h [hunk 2]
```diff
@@ -180,8 +186,10 @@ void ScannerControllerT<TCSM, TUCI>::handleScannerReply(const MaxSizeRawData& da
 template <typename TCSM, typename TUCI>
 void ScannerControllerT<TCSM, TUCI>::handleStartReplyTimeout(const std::string& error_str)
 {
-  PSENSCAN_ERROR(
-      "ScannerController", "Timeout while waiting for start reply message from scanner | Error message: {}", error_str);
+  PSENSCAN_ERROR("ScannerController",
+                 "Timeout while waiting for the scanner to start! Retrying..."
+                 "(Eventually check your setup or contact PILZ support)");
```
Comment (SansoneG) [inline include/psen_scan_v2/scanner_controller.h:191]:
```suggestion
                 "(Please check the ethernet connection or contact PILZ support if the error persists)");
```

File: include/psen_scan_v2/udp_connection_state_machine.h [hunk 3]
```diff
@@ -193,27 +224,31 @@ struct udp_connection_state_machine_ : public msm::front::state_machine_def<udp_
...
+  void no_transition(Event const& event, FSM&, int state)
+  {
+    PSENSCAN_WARN("StateMachine", "No transition in state {} for event {}.", state, classNameShort(event));
+  }
+
+  template <class FSM>
+  void no_transition(e::monitoring_frame_received const& event, FSM&, int state)
   {
-    // TODO Implement handling
+    PSENSCAN_WARN("StateMachine", "Received monitoring frame despite not waiting for it");
```
Comment (SansoneG) [inline include/psen_scan_v2/udp_connection_state_machine.h:250]: Could add in which state the monitoring frame was receive despite not waiting for it.

File: test/unit_tests/unittest_scanner_controller.cpp [hunk 4]
```diff
@@ -134,13 +134,46 @@ TEST_F(ScannerControllerTest, testSuccessfulStartSequence)
...
+
+TEST_F(ScannerControllerTest, testReceivingMultipleStartReplies)
+{
+  const UDPFrameTestDataWithoutIntensities test_data;
+  const LaserScan scan{ testDataToLaserScan(test_data) };
+
+  EXPECT_CALL(mock_, laserscan_callback(scan)).Times(2);
 
   scanner_controller_.start();
   simulateUdpTimeout("Udp timeout");
```
Comment (SansoneG) [inline test/unit_tests/unittest_scanner_controller.cpp:172]: Why is this timeout necessary?

File: test/unit_tests/unittest_controller_state_machine.cpp [hunk 5]
```diff
@@ -29,55 +44,97 @@ namespace psen_scan_v2_test
...
-                            std::bind(&ControllerStateMachineTest::started_cb, this),
-                            std::bind(&ControllerStateMachineTest::stopped_cb, this));
-
-  sm.processStartRequestEvent();
-  sm.processReplyReceivedEvent(ScannerReplyMsgType::Start);
-  sm.processStopRequestEvent();
-  sm.processReplyReceivedEvent(ScannerReplyMsgType::Stop);
+  EXPECT_CALL(controller_, start_request_cb()).Times(1);
+  state_machine_->processStartRequestEvent();
+  state_machine_->processStartRequestEvent();
```
[the remark this answers is not in the document]
Comment (rfeistenauer) [reply test/unit_tests/unittest_controller_state_machine.cpp]: No, only Debug message. We don't test for debug messages so far which IMHO is good. So I tested it like this.

File: include/psen_scan_v2/scanner_controller.h [hunk 6]
```diff
@@ -50,6 +51,44 @@ static constexpr uint32_t DEFAULT_SEQ_NUMBER{ 0 };
...
+using namespace std::chrono_literals;
+class TimeoutTimer
+{
+public:
+  TimeoutTimer(const std::chrono::high_resolution_clock::duration& timeout,
+               const std::function<void()>& timeout_handler,
+               const std::chrono::high_resolution_clock::duration& start_timeout = 10ms)
+    : timer_thread_([this, timeout, timeout_handler]() {
+      thread_startetd_barrier_.release();
+      while (!terminated_ && !barrier_.waitTillRelease(timeout))
```
Comment (martiniil) [inline include/psen_scan_v2/scanner_controller.h:63]: I think the variable `terminate_` is superfluous.

File: include/psen_scan_v2/scanner_controller.h [hunk 7]
```diff
@@ -50,6 +51,44 @@ static constexpr uint32_t DEFAULT_SEQ_NUMBER{ 0 };
...
+      {
+        timeout_handler();
+      }
+    })
+  {
+    if (!thread_startetd_barrier_.waitTillRelease(start_timeout))
```
Comment (martiniil) [inline include/psen_scan_v2/scanner_controller.h:69]: Which problem is solved by waiting for the thread to start?

File: include/psen_scan_v2/scanner_controller.h (hunk 6 above)
Comment (hslusarek) [reply include/psen_scan_v2/scanner_controller.h:63]: :smile: Yes, thx.

File: include/psen_scan_v2/scanner_controller.h (hunk 7 above)
Comment (hslusarek) [reply include/psen_scan_v2/scanner_controller.h:69]: To ensures that the timer is actually started, before we do anything else.

File: include/psen_scan_v2/scanner_controller.h (hunk 6 above)
Comment (hslusarek) [reply include/psen_scan_v2/scanner_controller.h]: Done.

File: include/psen_scan_v2/scanner_controller.h (hunk 7 above)
Comment (hslusarek) [reply include/psen_scan_v2/scanner_controller.h:73]: Done.

File: test/unit_tests/unittest_scanner_controller.cpp (hunk 4 above)
Comment (hslusarek) [reply test/unit_tests/unittest_scanner_controller.cpp]: Deleted -> Done.

**Pull request merged**